### PR TITLE
Rename `serve` to `run`, add asynchronous `serve`

### PR DESCRIPTION
**Pull request by tirr-c**
## Description
* Rename original `serve` to `run`.
* Add `App::serve` that returns a `Future`.

## Motivation and Context
This provides flexibility when composing app server with other futures. (cc #163)

## How Has This Been Tested?
No new tests, existing tests pass.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/rustasync/tide/blob/master/.github/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by fairingrey** [comment]
LGTM! Will likely have to be rebased after https://github.com/rustasync/tide/pull/219 lands, but shouldn't take long.

Also resolved the merge conflict (accepted the incoming changes), since `serve` now, *ahem*, serves a different purpose.

File: src/app.rs [hunk 1]
```diff
@@ -246,6 +246,19 @@ impl<State: Send + Sync + 'static> App<State> {
...
+    /// Asynchronously serve the app at the given address.
+    #[cfg(feature = "hyper")]
+    pub async fn serve(self, addr: impl std::net::ToSocketAddrs) -> std::io::Result<()> {
+        let addr = addr
+            .to_socket_addrs()?
+            .next()
+            .ok_or(std::io::ErrorKind::InvalidInput)?;
+
+        // TODO: propagate the error from hyper
+        await!(http_service_hyper::serve(self.into_http_service(), addr)).ok();
```
Comment (fairingrey) [inline src/app.rs]: This will need to use the new await syntax:

```rust
http_service_hyper::serve(self.into_http_service(), addr).await.ok();
```

Comment (tirr-c) [reply src/app.rs]: Done!

**Comment by tirr-c** [comment]
Rebased and review applied!

**Pull request merged**